files/image_config/hostcfgd/hostcfgd

```diff
@@ -219,6 +220,442 @@ class AaaCfg(object):
+        cfg_attr_set(cmd, 'speed', speed, appIntf)
+
+    def mgmt_intf_autoneg_set (self, ifname, autoneg, appIntf):
+        an = "on"
+        if autoneg == "true":
+            an = "on"
+        elif autoneg == "false":
+            an = "off"
+
+        cmd = "ethtool -s {} autoneg {}".format(ifname, an)
+        cfg_attr_set(cmd, 'autoneg', autoneg, appIntf)
+
+    def mgmt_intf_gwaddr_set(self, ifname, ip_prefix, gwaddr, op):
+        vrf_table = self.mgmt_vrf_cfg.mgmt_vrf_table_get()
+        ip_conf = netaddr.IPNetwork(ip_prefix)
+        ver = "-4"
+        if ip_conf.version == 6:
+            ver = "-6"
+
+        cmd = "ip {} route {} default via {} dev {} table {}".format(ver, op.lower(), gwaddr, ifname, vrf_table)
+        cfg_attr_set(cmd)
+
+
+    def mgmt_intf_dhclient_set (self, ifname, attr, enable):
+        version  = ""
+        file_ext = ""
+        cmd_opt  = ""
+
+        if attr == "ipv6_dhcp_client":
+            version = "-6"
+            file_ext = "6"
+            cmd_opt = "-D LL"
+
+        if enable == True:
+            path = "/var/run/dhclient{}.{}.pid".format(file_ext, ifname)
+            if os.path.exists(path):
+                cmd = "[ -f /var/run/dhclient{}.{}.pid ] && kill `cat /var/run/dhclient{}.{}.pid` && rm -f /var/run/dhclient{}.{}.pid".format(file_ext, ifname, file_ext, ifname, file_ext, ifname)
+                run_command(cmd)
+
+        if enable == True:
+            cmd = "/sbin/dhclient {} -pf /run/dhclient{}.{}.pid -lf /var/lib/dhcp/dhclient{}.{}.leases {} -nw {} ".format(version, file_ext, ifname, file_ext, ifname, ifname, cmd_opt)
+        else:
+            cmd = "/sbin/dhclient {} -r {} && [ -f /var/run/dhclient{}.{}.pid ] && kill `cat /var/run/dhclient{}.{}.pid` && rm -f /var/run/dhclient{}.{}.pid".format(version, ifname, file_ext, ifname, file_ext, ifname, file_ext, ifname)
+
+        run_command(cmd)
+
+    def mgmt_intf_ip_prefix_set (self, ifname, ip_prefix, op):
+        ver = "-4"
+        mgmt_conf = netaddr.IPNetwork(ip_prefix)
+        if mgmt_conf.version == 6:
+            ver = "-6"
+
+        cmd = "ip {} addr {} {} dev {}".format(ver, op.lower(), ip_prefix, ifname)
+        cfg_attr_set(cmd)
+
+        cmd = "ip {} rule {} from {} table default".format(ver, op.lower(), str(mgmt_conf.ip))
+        cfg_attr_set(cmd)
+
+    def mgmt_intf_forced_mgmt_routes_set (self, ifname, ip_prefix, forced_routes, op, app_forced_routes = []):
+        vrf_table = self.mgmt_vrf_cfg.mgmt_vrf_table_get()
+        if op == "UPDATE":
+            routes_add = []
+            routes_del = []
+            for cfg_route in forced_routes:
+                if cfg_route not in app_forced_routes:
+                    routes_add.append(cfg_route)
+
+            for cfg_route in app_forced_routes:
+                if cfg_route not in forced_routes:
+                    routes_del.append(cfg_route)
+
+            self.mgmt_intf_forced_mgmt_routes_set(ifname, ip_prefix, routes_del, "DEL")
```

Comment:
Is this a recursive function? What is achieved here?

Reply:
This function handle 3 operations ADD, DEL and UPDATE. For add and delete operations we can just loop through leaf list and perform of the operations. But in case of UPDATE operation as its a leaf list we  need to find which routes need to be deleted and which needs to be added between old and current configuration and perform ADD and DELETE operations.